File: Capítulo_14/Programação dinâmica/Multiplicação de cadeia de matrizes/matrixChaiMultiplication.js
// Algoritmo de multiplicação de cadeia de matrizes
function matrixChainOrder(p) {
    const n = p.length;
    const m = [];
    const s = [];
    for (let i = 1; i <= n; i++) {
        m[i] = [];
        m[i][i] = 0;
    }
    for (let i = 0; i <= n; i++) {
        s[i] = [];
        for (let j = 0; j <= n; j++) {
            s[i][j] = 0;
        }
    }
    for (let l = 2; l < n; l++) {
        for (let i = 1; i <= n - l + 1; i++) {
            const j = i + l - 1;
            m[i][j] = Number.MAX_SAFE_INTEGER;
            for (let k = i; k <= j - 1; k++) {
                const q = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j]; // q = cost/scalar multiplications
                if (q < m[i][j]) {
                    m[i][j] = q;
                    s[i][j] = k; // s[i,j] = Second auxiliary table that stores k
                }
            }
        }
    }
    printOptimalParenthesis(s, 1, n - 1);
    return m[1][n - 1];
}

// Função printOptimalParenthesis
function printOptimalParenthesis(s, i, j) {
    let result = "";

    function buildOptimalParenthesis(s, i, j) {
        if (i === j) {
            // console.log("A[" + i + "]");
            result += "A[" + i + "]";
        } else {
            // console.log("(");
            result += "(";
            buildOptimalParenthesis(s, i, s[i][j]);
            buildOptimalParenthesis(s, s[i][j] + 1, j);
            //console.log(")");
            result += ")";
        }
    }
    buildOptimalParenthesis(s, i, j);
    console.log(result)
}

// Testando o algoritmo matrixChainOrder
const p = [10, 100, 5, 50, 1];
const minCost = matrixChainOrder(p);
console.log(`O custo mínimo de multiplicação da cadeia de matrizes é: ${minCost}`);

File: Capítulo_14/Programação dinâmica/Multiplicação de cadeia de matrizes/Códigos em Python/matrixChaiMultiplication.py
# Algoritmo de multiplicação de cadeia de matrizes matrix_chain_order:
def matrix_chain_order(p):
    n = len(p)
    m = [[0 for _ in range(n)] for _ in range(n + 1)] # Inicialização da matrix m
    s = [[0 for _ in range(n)] for _ in range(n + 1)] # Inicialização da matrix s

    for l in range(2, n):
        for i in range(1, (n - l) + 1):
            j = (i + l) -1
            m[i][j] = float("inf")
            
            for k in range(i, j):
                q = m[i][k] + m[k + 1][j] + ((p[i - 1] * p[k]) * p[j])# q = cost/scalar multiplications
                if q < m[i][j]:
                    m[i][j] = q 
                    s[i][j] = k # s[i,j] = Second auxiliary table that stores k

    print_optimal_parenthesis(s, 1, n - 1)
    return m[i][n - 1]

# Função print_optimal_parenthesis
def  print_optimal_parenthesis(s, i, j):
    result = ""

    def build_optimal_parenthesis(s, i, j):
        nonlocal result  # Permite modificar a variável `result` do escopo externo
        if i == j:
            # print("A[", i, "]")
            result += "A[" + str(i) + "]"
        else:
            # print("(")
            result += "("
            build_optimal_parenthesis(s, i, s[i][j])
            build_optimal_parenthesis(s, s[i][j] + 1, j)
            # print(")")
            result += ")"

    build_optimal_parenthesis(s, i, j)
    print(result)

# Testando o algoritmo matrixChainOrder
p = [10, 100, 5, 50, 1]
minCost = matrix_chain_order(p)
print(f"O custo mínimo de multiplicação da cadeia de matrizes é: {minCost}")
